src/cmap/worker.rs

```diff
@@ -0,0 +1,494 @@
+                        &establisher,
+                        pending_connection,
+                        &manager,
+                        event_handler.as_ref(),
+                    )
+                    .await;
+
+                    if let Ok(connection) = connection {
+                        manager.populate_connection(connection);
+                    }
+                });
+            }
+        }
+    }
+}
+
+async fn establish_connection(
+    establisher: &ConnectionEstablisher,
+    pending_connection: PendingConnection,
+    manager: &PoolManager,
+    event_handler: Option<&Arc<dyn CmapEventHandler>>,
+) -> Result<Connection> {
+    let mut establish_result = establisher.establish_connection(pending_connection).await;
+
+    match establish_result {
+        Err(ref e) => {
+            manager.handle_connection_failed(e.clone());
+        }
+        Ok(ref mut connection) => {
+            if let Some(handler) = event_handler {
+                handler.handle_connection_ready_event(connection.ready_event())
+            };
+        }
+    }
+
+    establish_result
+}
+
+/// Task to process by the worker.
+#[derive(Debug)]
+enum PoolTask {
+    /// Clear the pool.
+    Clear,
+
+    /// Check in the given connection.
+    CheckIn(Connection),
+
+    /// Fulfill the given connection request.
+    CheckOut(ConnectionRequest),
+
+    /// Add this new connection to the pool without emitting events.
+    Populate(Connection),
+
+    /// Update the pool state based on the given establishment error.
+    HandleConnectionFailed(Error),
+
+    /// Perform pool maintenance (ensure min connections, remove stale or idle connections).
+    Maintenance,
+}
+
+impl From<PoolManagementRequest> for PoolTask {
+    fn from(request: PoolManagementRequest) -> Self {
+        match request {
+            PoolManagementRequest::CheckIn(c) => PoolTask::CheckIn(c),
+            PoolManagementRequest::Clear => PoolTask::Clear,
+            PoolManagementRequest::HandleConnectionFailed(error) => {
+                PoolTask::HandleConnectionFailed(error)
+            }
+            PoolManagementRequest::Populate(c) => PoolTask::Populate(c),
+        }
+    }
+}
+
+/// Handle to the worker. Once all handles have been dropped, the worker
+/// will stop waiting for new requests and drop the pool itself.
+#[derive(Debug, Clone)]
+pub(super) struct PoolWorkerHandle {
+    sender: mpsc::Sender<()>,
+}
+
+/// Listener used to determine when all handles have been dropped.
+#[derive(Debug)]
+struct HandleListener {
+    receiver: mpsc::Receiver<()>,
+}
+
+impl HandleListener {
+    fn new() -> (Self, PoolWorkerHandle) {
+        let (sender, receiver) = mpsc::channel(1);
+        (Self { receiver }, PoolWorkerHandle { sender })
+    }
+
+    /// Listen until all handles are dropped.
+    /// This will not return until all handles are dropped, so make sure to only poll this via
+    /// select or with a timeout.
+    async fn listen(&mut self) {
```

Comment:
Could we name something that conveys its usage a bit more clearly, e.g. "wait_for_handle_drop"?

Reply:
Yep, that's perfect